internal/wasm/jit/jit_arm64.go

```diff
@@ -1157,8 +1344,103 @@ func (c *arm64Compiler) compileReadInstructionAddress(beforeTargetInst obj.As, d
 	})
 }
 
+// compileCallIndirect implements compiler.compileCallIndirect for the arm64 architecture.
 func (c *arm64Compiler) compileCallIndirect(o *wazeroir.OperationCallIndirect) error {
-	return fmt.Errorf("TODO: unsupported on arm64")
+	offset := c.locationStack.pop()
+	if err := c.compileEnsureOnGeneralPurposeRegister(offset); err != nil {
+		return err
+	}
+
+	if isZeroRegister(offset.register) {
+		reg, err := c.allocateRegister(generalPurposeRegisterTypeInt)
+		if err != nil {
+			return err
+		}
+		offset.setRegister(reg)
+		c.markRegisterUsed(reg)
+
+		// Zero the value on a picked register.
+		c.compileRegisterToRegisterInstruction(arm64.AMOVD, zeroRegister, reg)
+	}
+
+	tmp, err := c.allocateRegister(generalPurposeRegisterTypeInt)
+	if err != nil {
+		return err
+	}
+
+	// First, we need to check if the offset doesn't exceed the length of table.
+	// "tmp = len(table)"
+	c.compileMemoryToRegisterInstruction(arm64.AMOVD,
+		reservedRegisterForEngine, engineModuleContextTableSliceLenOffset,
+		tmp,
+	)
+	// "cmp tmp, offset"
+	c.compileTwoRegistersToNoneInstruction(arm64.ACMP, tmp, offset.register)
+
+	// If it exceeds len(table), we exit the execution.
+	brIfOffsetOK := c.compilelBranchInstruction(arm64.ABLO)
+	c.compileExitFromNativeCode(jitCallStatusCodeInvalidTableAccess)
+
+	// Otherwise, we proceed to do function type check.
+	c.setBranchTargetOnNext(brIfOffsetOK)
+
+	// We need to obtains the absolute address of table element.
+	// "tmp = &table[0]"
+	c.compileMemoryToRegisterInstruction(
+		arm64.AMOVD,
+		reservedRegisterForEngine, engineModuleContextTableElement0AddressOffset,
+		tmp,
+	)
+	// "offset = tmp + (offset << 4) (== &table[offset])"
+	c.compileAddInstructionWithLeftShiftedRegister(
+		offset.register, 4,
+		tmp,
+		offset.register,
+	)
+
+	// Check if table[offset].TypeID == targetFunctionType.
+	targetFunctionType := c.f.ModuleInstance.Types[o.TypeIndex]
+	// "tmp = table[offset].TypeID"
+	c.compileMemoryToRegisterInstruction(
+		arm64.AMOVD, offset.register, tableElementFunctionTypeIDOffset,
+		tmp,
+	)
+	// "reservedRegisterForTemporary = targetFunctionType.TypeID"
+	c.compileConstToRegisterInstruction(arm64.AMOVD, int64(targetFunctionType.TypeID), reservedRegisterForTemporary)
+	// Compare these two values, and if they equal, we are ready to make function call.
+	c.compileTwoRegistersToNoneInstruction(arm64.ACMP, tmp, reservedRegisterForTemporary)
+	brIfTypeMatched := c.compilelBranchInstruction(arm64.ABEQ)
+
+	// Otherwise, we have to exit the execution with either jitCallStatusCodeTypeMismatchOnIndirectCall or jitCallStatusCodeInvalidTableAccess.
```

Comment:
the bracing is nice to understand and I guess needed as this amount of statements wouldn't be inlined by go?

Reply:
I just have a brace here for readability, in anyway the statements are inlined